### PR TITLE
refactor(component-mapper): introduce a proxy for module.exports

### LINKED ISSUE #885: CompartmentMapper - cjs exports defined through defineProperty are not handled correctly
example defineProperty based exports in the wild https://npmfs.com/package/ethereumjs-util/7.1.0/dist/index.js#L4

in this case this problem might be worked around by convincing the package author to also distribute esm format

### PR DESCRIPTION
**Pull request by naugtur**
Fix #885 

I don't yet know if it's ok to leave Object as the target for the proxy and return its prototype methods when requested.   
CommonJS lets modules depend on `exports` being an object with the default prototype, so it needs to be there. Maybe hardened or petrified? 

I had this code for a few days now, but wanted to see if I can use the proxy to fix `import * from cjs`. Turns out it has nothing to do with parse-cjs so this can land separately.

### CODE REVIEW

File: packages/compartment-mapper/test/fixtures-cjs-compat/node_modules/defineprop/main.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    o[k2] = m[k];
+}));
+var __exportStar = (this && this.__exportStar) || function(m, exports) {
+    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
+};
+Object.defineProperty(exports, "__esModule", { value: true });
+/// <reference path="./@types/ethjs-util/index.ts"/>
+
+const stuff = { stuff: 1 };
+__exportStar(stuff, exports);
```
Comment (kriskowal) [inline packages/compartment-mapper/test/fixtures-cjs-compat/node_modules/defineprop/main.js:17]: Should have some assertion on the import side.

**Review by kriskowal — CHANGES_REQUESTED** [review]
Great test. Some suggestions.

File: packages/compartment-mapper/src/parse-cjs.js [hunk 2]
```diff
@@ -29,7 +29,22 @@ export const parseCjs = async (
     const functor = compartment.evaluate(
       `(function (require, exports, module, __filename, __dirname) { ${source} //*/\n})\n//# sourceURL=${location}`,
     );
-    let exportsReferenceCopy = moduleExports;
+
+    const exportsDoubleProxy = new Proxy(
+      {},
+      {
+        get(target, prop) {
+          return moduleExports[prop] || target[prop]; // this makes things like exports.hasOwnProperty() work. Does it break hardening though?
```
[the remark this answers is not in the document]
Comment (naugtur) [reply packages/compartment-mapper/src/parse-cjs.js:37]: I assumed moduleExports were already a good name but did count on hints for the double proxy. I only named it that knowing moduleExports was already acting as a proxy - meant as a hint to the layers of proxy underneath.

I get the .freeze part - should have applied that to begin with.
What I can't wrap m head around right now is if it's ok to have the target of the proxy be an object with the actual prototype and make the prototype accessible through the `|| target[prop];`  
It gives me support for OBject methods, but it seems like it's the wrong prototype to expose here.

Can I get Object.prototype from the compartment here? Should I try?

File: packages/compartment-mapper/test/fixtures-cjs-compat/node_modules/defineprop/main.js (hunk 1 above)
Comment (naugtur) [reply packages/compartment-mapper/test/fixtures-cjs-compat/node_modules/defineprop/main.js:17]: :+1:

File: packages/compartment-mapper/src/parse-cjs.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (naugtur) [reply packages/compartment-mapper/src/parse-cjs.js]: @kriskowal Let's discuss this so the PR can be wrapped up, 
I've got a chain of other PRs depending on it :)

**Review by kriskowal — APPROVED** [review]

**Pull request merged**